# SARS-CoV-2-Indels
This repository includes data and codes we used to study insersions and deletions (indels) in the SARS-CoV-2 genomes. 

We show that in-frame indels (IF-indels) play a significant and rapidly increasing role in the evolution of the SARS-CoV-2 genome. The percentage of genomes and lineages with IF-indels is increasing rapidly, especially in the emerging variants of concern (VOCs). Monitoring of SARS-CoV-2 genome evolution uncovers co-occurrence (i.e., highly correlated presence) of IF-indels, especially IF-indels on spike N-terminal domain and non-structural protein 6 (NSP6) is a shared feature in several VOCs such as Alpha, Beta, Delta, and Omicron. IF-indels are highly recurrent (appear independently multiple times) in SARS-CoV-2 proteins and appear to affect interactions with the host immune components. IF-indels distribution is correlated with spike mutations associated with immune escape and growth in the number of genomes with IFDs coincides with the increasing population resistance due to vaccination and previous infections. The IF-indels occur in the spike glycoprotein, but also in other proteins, especially those involved in interactions with the host immune system. We also showed that IF-indels concentrated on specific regions of individual SARS-CoV-2 proteins known as recurrent-deletion regions (RDRs) or recurrent-insertion regions (RIRs), which are mostly located in protein loop regions. Structural analysis suggests that IF-indels remodel viral proteins’ surfaces at common epitopes and interaction interfaces, affecting the virus’ interactions with host proteins. We hypothesize that the increased frequency of IF-indels, the non-random distribution of IF-indels and their independent co-occurrence in several VOCs from diverse geographic origins (i.e., repeated appearance of the same combinations of IF-indels) is an adaptive response to elevated global population immunity.

Cite Us
